src/material-experimental/mdc-checkbox/checkbox.ts

```diff
@@ -0,0 +1,296 @@
+    // ripple, which we do ourselves instead.
+  }
+
+  ngAfterViewInit() {
+    this._rippleTarget = {
+      rippleConfig: {
+        radius: 20,
+        centered: true,
+        animation: {
+          // TODO(mmalerba): Use the MDC constants once they are exported separately from the
+          // foundation. Grabbing them off the foundation prevents the foundation class from being
+          // tree-shaken.
+          enterDuration: 225 /* MDCRippleFoundation.numbers.DEACTIVATION_TIMEOUT_MS */,
+          exitDuration: 150 /* MDCRippleFoundation.numbers.FG_DEACTIVATION_MS */,
+        },
+      },
+      rippleDisabled: true
+    };
+    this._rippleRenderer =
+        new RippleRenderer(this._rippleTarget, this._ngZone, this._checkbox, this._platform);
+
+    this._checkboxFoundation.init();
+  }
+
+  ngOnDestroy() {
+    this._checkboxFoundation.destroy();
+  }
+
+  registerOnChange(fn: (checked: boolean) => void) {
+    this._cvaOnChange = fn;
+  }
+
+  registerOnTouched(fn: () => void) {
+    this._cvaOnTouch = fn;
+  }
+
+  setDisabledState(isDisabled: boolean) {
+    this.disabled = isDisabled;
+  }
+
+  writeValue(value: any) {
+    this.checked = !!value;
+    this._cdr.markForCheck();
+  }
+
+  focus() {
+    this._nativeCheckbox.nativeElement.focus();
+  }
+
+  toggle() {
+    this.checked = !this.checked;
+  }
+
+  _activateRipple() {
+    if (!this.disabled && !this.disableRipple && this._animationMode != 'NoopAnimations') {
+      this._rippleRenderer.fadeInRipple(0, 0, this._rippleTarget.rippleConfig);
+    }
+  }
+
+  _onBlur() {
+    Promise.resolve().then(() => {
+      this._cvaOnTouch();
+      this._cdr.markForCheck();
+    });
+  }
+
+  // Note: we must listen to the `click` event rather than the `change` event because IE & Edge do
+  // not fire a `change` event when the user clicks on an indeterminate checkbox.
+  _onClick() {
+    if (this._clickAction === 'noop') {
+      this._nativeCheckbox.nativeElement.checked = this.checked;
+      this._nativeCheckbox.nativeElement.indeterminate = this.indeterminate;
+      return;
+    }
+
+    if (this.indeterminate && this._clickAction !== 'check') {
+      this.indeterminate = false;
+      Promise.resolve().then(() => this.indeterminateChange.next(this.indeterminate));
+    } else {
+      this._nativeCheckbox.nativeElement.indeterminate = this.indeterminate;
+    }
+
+    this.checked = this._nativeCheckbox.nativeElement.checked;
+    this._checkboxFoundation.handleChange();
+
+    // Dispatch our change event
+    const newEvent = new MatCheckboxChange();
+    newEvent.source = this as any;
+    newEvent.checked = this.checked;
+    this._cvaOnChange(this.checked);
+    this.change.next(newEvent);
+  }
+
+  _getAriaChecked(): 'true'|'false'|'mixed' {
+    return this.checked ? 'true' : (this.indeterminate ? 'mixed' : 'false');
+  }
+
+  private _setClass(cssClass: string, active: boolean) {
```

Comment:
are are you going to make this shared for other components to use? add/removeClass is probably in 90% of the components with an adapter.

Reply:
We can look at doing a mixin if we see multiple components having the exact same behavior